lib/Parse/ParseDecl.cpp

```diff
@@ -2926,9 +2926,25 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
     if (!Tok.is(tok::l_paren)) {
       // Normal access control attribute.
       AttrRange = Loc;
-      DuplicateAttribute = Attributes.getAttribute<AccessControlAttr>();
-      if (!DuplicateAttribute)
+      
+      auto PrevAccessControlAttr = Attributes.getAttribute<AccessControlAttr>();
+      
+      if (!PrevAccessControlAttr) {
         Attributes.add(new (Context) AccessControlAttr(AtLoc, Loc, access));
+      } else {
+        // Diagnose if there is already an access control attribute on
+        // this declaration.
+        diagnose(Loc, diag::multiple_access_level_modifiers)
+          .highlight(AttrRange);
+        diagnose(PrevAccessControlAttr->getLocation(),
+                 diag::previous_access_level_modifier,
+                 PrevAccessControlAttr->getAttrName())
+          .highlight(PrevAccessControlAttr->getRange());
+        
+        // Removes the reference to the duplicate attribute as we don't need the
+        // extra diagnostic.
+        DuplicateAttribute = nullptr;
```

Comment:
I think this is redundant because you deleted the line that was setting it for access-level modifiers.

Reply:
It's not redundant because this variable is set on the beginning of the function body. If I keep it set, it will trigger the original "Duplicate modifier" diagnostic.

Reply:
Oh, I see, `DuplicateAttribute` is supposed to be a constant. Could you adjust the logic to use the `DiscardAttribute` flag, as other cases do, and remove the mutations?

Reply:
At the beginning of the `case` there's one reassignment that looks unnecessary and I removed it now, as requested.

But I'm not sure `DuplicateAttribute` is supposed to be a constant. The original logic reassigns it to check for attributes like `private(set)`, that wouldn't be checked otherwise. So, I kept and improved it by writing a comment and updating the `DiscardAttribute` boolean, that was becoming outdated after the reassignment. Let me know what you think about it.